source/Engine.cpp

```diff
@@ -870,9 +866,29 @@ void Engine::Step(bool isActive)
 	// Draw crosshairs on any minables in range of the flagship's scanners.
 	double scanRange = flagship ? 100. * sqrt(flagship->Attributes().Get("asteroid scan power")) : 0.;
 	if(flagship && scanRange && !flagship->IsHyperspacing())
+	{
+		// Decide before looping whether or not to catalog asteroids.  This
+		// results in cataloging in-range asteroids roughly 3 times a second.
+		bool shouldCatalogAsteroids = !Random::Int(20);
+		shouldCatalogAsteroids &= !isAsteroidCatalogComplete;
+		bool scanComplete = true;
 		for(const shared_ptr<Minable> &minable : asteroids.Minables())
 		{
 			Point offset = minable->Position() - center;
+
+			// Autocatalog asteroid: Record that the player knows this type of asteroid is available here.
+			if(shouldCatalogAsteroids)
+			{
+				if(!Random::Int(10) && (minable->Position() - flagship->Position()).Length() <= scanRange)
```

Comment:
We can do minable->Position().DistanceSquared(flagship->Position()) for optimisation, and drop the sqrt earlier (867)
Just don't multiply it by 100 but by 10.

Reply:
Also, I just noticed the !asteroidsScanned.count lower.
We should definitely be putting that condition in front of this, and have this be the else, if you see what I mean?

if(!asteroidsScanned.count(minable->Name()))
					scanComplete = false;
else
{
if(random::Int(10).................
}

Reply:
It's intentionally lower in case player.Harvest is called on the current asteroid and that's the last one to be harvested.  Otherwise it would go through another loop.  In practice probably not a big deal ordered one way or the other.

Reply:
I will move it up and combine your feedback with quyyk's